src/components/membership-queue.ts

```diff
@@ -279,15 +280,17 @@ export class MembershipQueue {
         }
     }
 
-    private shouldRetry(ex: {body: {code: string; errcode: string;}, statusCode: number}, attempts: number): boolean {
-        const { errcode } = ex.body;
-        return !(
-            attempts === this.opts.maxAttempts ||
-            // Forbidden
-            errcode === "M_FORBIDDEN" ||
-            ex.statusCode === 403 ||
-            // Not found
-            ex.statusCode === 404
-        );
+    private shouldRetry(ex: Error, attempts: number): boolean {
+        if (ex instanceof MatrixError) {
+            return !(
+                attempts === this.opts.maxAttempts ||
+                // Forbidden
+                ex.errcode === "M_FORBIDDEN" ||
+                ex.statusCode === 403 ||
+                // Not found
+                ex.statusCode === 404
+            )
+        }
+        return false;
```

Comment:
Should this return `true` instead?  If `ex` isn't a `MatrixError`, the original code can resolve to `!(undefined === <something>)` -> `!false` -> `true`